### PR TITLE
[PM-2133] Improve channel closing in tls peer group

### PR DESCRIPTION
**Pull request by KonradStaniec**

### CODE REVIEW

File: scalanet/src/io/iohk/scalanet/peergroup/dynamictls/DynamicTLSPeerGroupInternals.scala [hunk 1]
```diff
@@ -196,10 +188,9 @@ private[dynamictls] object DynamicTLSPeerGroupInternals {
     override def close(): Task[Unit] = {
       for {
         _ <- Task.now(log.debug("Closing client channel to peer {}", peerInfo))
-        _ <- Task(messageSubject.onComplete())
         ctx <- Task.fromFuture(activationF)
         _ <- toTask(ctx.close())
-        _ <- Task.fromFuture(deactivationF)
+        _ <- toTask(ctx.closeFuture())
```
Comment (ntallar) [inline scalanet/src/io/iohk/scalanet/peergroup/dynamictls/DynamicTLSPeerGroupInternals.scala:197]: Out of lack of knowledge, why are both close needed here?

File: scalanet/src/io/iohk/scalanet/peergroup/dynamictls/DynamicTLSPeerGroupInternals.scala [hunk 2]
```diff
@@ -246,6 +237,11 @@ private[dynamictls] object DynamicTLSPeerGroupInternals {
                       .HandshakeFailed(new HandshakeException(PeerInfo(BitVector.empty, remoteAddress), e.cause()))
                   )
               }
+            case _ =>
+              log.info(
```
Comment (ntallar) [inline scalanet/src/io/iohk/scalanet/peergroup/dynamictls/DynamicTLSPeerGroupInternals.scala]: Should this be a debug log?

File: scalanet/src/io/iohk/scalanet/peergroup/dynamictls/DynamicTLSPeerGroupInternals.scala [hunk 3]
```diff
@@ -281,9 +277,9 @@ private[dynamictls] object DynamicTLSPeerGroupInternals {
 
     override def close(): Task[Unit] =
       for {
-        _ <- Task.now(log.debug("Closing client channel to peer {}", to))
-        _ <- Task(messageSubject.onComplete())
```
Comment (ntallar) [inline scalanet/src/io/iohk/scalanet/peergroup/dynamictls/DynamicTLSPeerGroupInternals.scala:285]: Why is this no longer needed?

**Review by ntallar — CHANGES_REQUESTED** [review]

File: scalanet/src/io/iohk/scalanet/peergroup/dynamictls/DynamicTLSPeerGroupInternals.scala (hunk 1 above)
Comment (KonradStaniec) [reply scalanet/src/io/iohk/scalanet/peergroup/dynamictls/DynamicTLSPeerGroupInternals.scala:197]: we we want to be sure that `channelnactive` handler had run before returning from task. In theory it should happen just with `toTask(ctx.close())` but to be sure we are are also waiting on `closeFuture` which is actually and promise.

File: scalanet/src/io/iohk/scalanet/peergroup/dynamictls/DynamicTLSPeerGroupInternals.scala (hunk 3 above)
Comment (KonradStaniec) [reply scalanet/src/io/iohk/scalanet/peergroup/dynamictls/DynamicTLSPeerGroupInternals.scala:285]: now `messageSubject.onComplete` is called only in `MessageNotifier.channelInactive` handler. This approach removes any possible violation of subject contract as all calls to message subject i.e`onNext`, `onComplete`, `onError` ale always happening on the same netty thread.

File: scalanet/src/io/iohk/scalanet/peergroup/dynamictls/DynamicTLSPeerGroupInternals.scala [hunk 4]
```diff
@@ -246,6 +237,11 @@ private[dynamictls] object DynamicTLSPeerGroupInternals {
                       .HandshakeFailed(new HandshakeException(PeerInfo(BitVector.empty, remoteAddress), e.cause()))
                   )
               }
+            case _ =>
+              log.debug(
+                s"User Event server channel from ${ctx.channel().localAddress()} " +
```
Comment (pslaski) [inline scalanet/src/io/iohk/scalanet/peergroup/dynamictls/DynamicTLSPeerGroupInternals.scala]: should we add also what kind of event we got?

**Review by pslaski — APPROVED** [review]

File: scalanet/src/io/iohk/scalanet/peergroup/dynamictls/DynamicTLSPeerGroupInternals.scala (hunk 1 above)
Comment (ntallar) [reply scalanet/src/io/iohk/scalanet/peergroup/dynamictls/DynamicTLSPeerGroupInternals.scala:197]: Should we add a comment on how this works maybe?

**Review by ntallar — APPROVED** [review]
LGTM!
```
     /(__M__)\
    /, ,   , ,\
   /' ' 'V' ' '\       ARV

------------------------------------------------
Thank you for visiting https://asciiart.website/
This ASCII pic can be found at
https://asciiart.website/index.php?art=animals/bats
```

**Pull request merged**